silhouette/app/com/mohiva/play/silhouette/api/util/PasswordHasher.scala

```diff
@@ -65,4 +62,71 @@ trait PasswordHasher {
    * @return True if the password matches, false otherwise.
    */
   def matches(passwordInfo: PasswordInfo, suppliedPassword: String): Boolean
+
+  /**
+   * Checks if a password hasher is suitable for the given password info.
+   *
+   * @param passwordInfo The password info to check the hasher is suitable for.
+   * @return True if the hasher is suitable for the given password info, false otherwise.
+   */
+  def isSuitable(passwordInfo: PasswordInfo): Boolean = passwordInfo.hasher == id
+
+  /**
+   * Indicates if a password info hashed with this hasher is deprecated.
+   *
+   * A password can be deprecated if some internal state of a hasher has changed.
+   *
+   * @param passwordInfo The password info to check the deprecation status for.
+   * @return True if the given password info is deprecated, false otherwise. If a hasher isn't
+   *         suitable for the given password, this method should return None.
+   */
+  def isDeprecated(passwordInfo: PasswordInfo): Option[Boolean]
+}
+
+/**
+ * Defines the password hashers used by the application.
+ *
+ * Sometimes it's needed to change the password hashing algorithm, because of a better algorithm or some
+ * similar case. But the passwords stored in the backing store cannot easily be converted with another
+ * algorithm because they're hashed and cannot be decrypted back to plain text. For such case Silhouette
+ * supports the change of hashing algorithms on the fly. So if a user successfully authenticates after
+ * the application has changed the hashing algorithm, the provider hashes the entered plain-text password
+ * again with the new algorithm and overrides the auth info in the backing store with the new hash.
+ *
+ * The password hasher registry defines the current password hasher which is able to hash all new passwords
+ * and also match the passwords store in the backing store for this algorithm. And also a list of deprecated
+ * hashers, which should match passwords that are stored in the baking store but which are different to the
+ * current hasher.
+ *
+ * @param current The current password hasher used by the application.
+ * @param deprecated The deprecated list of password hashers.
+ */
+case class PasswordHasherRegistry(current: PasswordHasher, deprecated: Seq[PasswordHasher] = Seq()) {
```

Comment:
Yes, it makes sense, because it defines the usage of the password hashers application wide. You must define it only once and all password based providers get it injected. If we remove this class, then every password based provider must define the list of hashers again. We have one table that stores the password info and we have one definition of the hashers stored in this table.